README.md
# pyright-polite

[![PyPi Version][pypi-img]][pypi-url]
[![License][license-img]][license-url]
[![Continuous Integration][ci-img]][ci-url]
[![Code Coverage][coverage-img]][coverage-url]
[![Python Versions][python-img]][python-url]

[pypi-img]: https://img.shields.io/pypi/v/pyright-polite.svg
[pypi-url]: https://pypi.org/project/pyright-polite
[license-img]:  https://img.shields.io/github/license/jamielinux/pyright-polite.svg
[license-url]: https://github.com/jamielinux/pyright-polite/blob/main/LICENSE
[ci-img]: https://github.com/jamielinux/pyright-polite/actions/workflows/ci.yml/badge.svg
[ci-url]: https://github.com/jamielinux/pyright-polite/actions/workflows/ci.yml
[coverage-img]: https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/jamielinux/ed2d4df7c2b137ac89778db60ef8894f/raw/pyright-polite.covbadge.json
[coverage-url]: https://github.com/jamielinux/pyright-polite/actions/workflows/ci.yml
[python-img]: https://img.shields.io/pypi/pyversions/pyright-polite.svg
[python-url]: https://pypi.org/project/pyright-polite

---

**pyright-polite** is an intelligent cross-platform wrapper for [pyright][0] that makes
it less noisy.

Force pyright to be more respectful with your attention :rotating_light:

[0]: https://github.com/microsoft/pyright

## What does it do?

With **pyright-polite**:

```console
$ pyright-polite
Found 7 source files
0 errors, 0 warnings, 0 informations
```

Without:

```console
$ pyright
WARNING: there is a new pyright version available (v1.1.300 -> v1.1.301).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

No configuration file found.
pyproject.toml file found at /projects/pyright_polite.
Loading pyproject.toml file at /projects/pyright_polite/pyproject.toml
Assuming Python version 3.11
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /projects/pyright_polite/typings is not a valid directory.
Searching for source files
Found 7 source files
pyright 1.1.300
0 errors, 0 warnings, 0 informations
Completed in 1.006sec
```

Now pyright is just as polite as your other tools:

```console
$ hatch run lint
cmd [1] | - ruff check .
cmd [2] | - black --quiet --check --diff .
cmd [3] | - pyright-polite
Found 8 source files
0 errors, 0 warnings, 0 informations
cmd [4] | - ssort --check --diff .
8 files would be left unchanged
```

Error messages are still shown (eg, if your pyright config file is invalid).

## Installation

You need `pyright` installed (ie, available somewhere in your `PATH`).

See pyright's installation instructions [here][installation]. Usually people install
either the [pyright npm][pkg_npm] or the [pyright PyPI][pkg_pypi] package.

```console
$ npm install pyright  # alternatively: pip install pyright
$ pip install pyright-polite
```

Linux, macOS and Windows are all supported.

[pkg_pypi]: https://pypi.org/project/pyright/
[pkg_npm]: https://www.npmjs.com/package/pyright
[installation]: https://microsoft.github.io/pyright/#/installation

## Usage

**pyright-polite** takes the same arguments as pyright.

```console
$ pyright-polite -h
Usage: pyright-polite [options] files...
  Options:
  --createstub <IMPORT>              Create type stub file(s) for import
  --dependencies                     Emit import dependency information
  -h,--help                          Show this help message
  --ignoreexternal                   Ignore external imports for --verifytypes
  --lib                              Use library code to infer types when stubs are missing
  --level <LEVEL>                    Minimum diagnostic level (error or warning)
  --outputjson                       Output results in JSON format
  -p,--project <FILE OR DIRECTORY>   Use the configuration file at this location
  --pythonplatform <PLATFORM>        Analyze for a specific platform (Darwin, Linux, Windows)
  --pythonversion <VERSION>          Analyze for a specific version (3.3, 3.4, etc.)
  --skipunannotated                  Skip analysis of functions with no type annotations
  --stats                            Print detailed performance stats
  -t,--typeshed-path <DIRECTORY>     Use typeshed type stubs at this location
  -v,--venv-path <DIRECTORY>         Directory that contains virtual environments
  --verbose                          Emit verbose diagnostics
  --verifytypes <PACKAGE>            Verify type completeness of a py.typed package
  --version                          Print Pyright version
  --warnings                         Use exit code of 1 if warnings are reported
  -w,--watch                         Continue to run and watch for changes


  Note: pyright-polite does not filter output from `--dependencies` or `--stats`.

```

## How does it work?

**pyright-polite** is hilariously over-engineered, but robust.

It aims to display everything in exactly the same format and colorisation as pyright
(minus the useless messages), and takes advantage of pyright's `--outputjson` option
when possible. It launches `pyright` as a subprocess and reads from both stderr and
stdout using `asyncio` tasks, which means that `--watch` is also supported.

For insight into what messages get hidden, see the [`print_filtered`][print_filtered]
method.

[print_filtered]: https://github.com/jamielinux/pyright-polite/blob/main/src/pyright_polite/pyright.py#L183-L233

## Isn't this a bit overkill?

Yes :rofl:

It was primarily a fun weekend project to learn `asyncio`.

## Why is pyright so noisy?

If you're wondering why `pyright` has to remind us that `typings is not a valid
directory` (among other useless messages) on literally every single launch, see
[pyright #4594][issue4594] for what the developers have to say:

> The current information output by the cli is there for a reason.

[issue4594]: https://github.com/microsoft/pyright/issues/4594

## License

`pyright-polite` is distributed under the terms of the [MIT][license] license.

[license]: https://spdx.org/licenses/MIT.html


CHANGELOG.md

<a id='changelog-v1.0.5'></a>
# v1.0.5 — 2023-04-11

## Removed

- Removed unnecessary `setup.cfg`.

## Added

- Added some more tests.

## Changed

- Tiny bit of code refactoring.

<a id='changelog-v1.0.4'></a>
# v1.0.4 — 2023-04-10

## Added

- Add `__slots__` to `Polite` class.

## Changed

- Tweak docstrings.

<a id='changelog-v1.0.3'></a>
# v1.0.3 — 2023-04-09

## Changed

- Small tweaks to improve readability.
- Improve README.

<a id='changelog-v1.0.2'></a>
# v1.0.2 — 2023-04-07

## Removed

- Exclude some files from source distribution.

<a id='changelog-v1.0.1'></a>
# v1.0.1 — 2023-04-06

## Changed

- Tweak pypi metadata.

<a id='changelog-v1.0.0'></a>
# v1.0.0 — 2023-04-06

## Added

- Initial release.


tests/test_cli.py
# SPDX-FileCopyrightText: Copyright © 2023 Jamie Nguyen <j@jamielinux.com>
# SPDX-License-Identifier: MIT

import sys
from unittest.mock import patch

import pytest
from pyright_polite.cli import Mode, parse_cli


def test_help(capsys):
    with patch("sys.argv", ["pyright-polite", "--help"]):
        with pytest.raises(SystemExit) as pytest_e:
            parse_cli()
        assert pytest_e.type == SystemExit
        assert pytest_e.value.code == 0
        captured = capsys.readouterr()
        assert captured.err == ""
        assert captured.out.startswith("Usage: pyright-polite [options] files...")


def test_no_args():
    with patch("sys.argv", ["pyright-polite"]):
        result = parse_cli()
        assert result.argv[1:] == ["--outputjson"]
        assert result.mode == Mode.JSON


@pytest.mark.parametrize(
    ("args", "mode"),
    [
        (["--version"], Mode.UNFILTERED),
        (["--createstub", "datetime"], Mode.PLAINTEXT),
        (["--dependencies"], Mode.UNFILTERED),
        (["--ignoreexternal", "--verifytypes", "foo"], Mode.PLAINTEXT),
        (["--stats"], Mode.UNFILTERED),
        (["--createstub", "datetime", "--stats"], Mode.UNFILTERED),
    ],
)
def test_not_json(args, mode):
    with patch("sys.argv", ["pyright-polite", *args]):
        result = parse_cli()
        assert result.argv[1:] == args
        assert result.mode == mode


@pytest.mark.parametrize(
    ("args"),
    [
        (["--lib"]),
        (["--level", "error"]),
        (["--project", "project"]),
        (["--pythonplatform", "Linux"]),
        (["--pythonversion", "3.11"]),
        (["--skipunannotated"]),
        (["--typeshed-path", "typeshed-path"]),
        (["--venv-path", "venv-path"]),
        (["--warnings"]),
        (["--verbose"]),
        (["--watch"]),
    ],
)
def test_json(args):
    with patch("sys.argv", ["pyright-polite", *args]):
        result = parse_cli()
        assert result.argv[1:] == [*args, "--outputjson"]
        assert result.mode == Mode.JSON


@pytest.mark.parametrize(
    ("args"),
    [
        (["--stats", "--outputjson"]),
    ],
)
def test_json_precedence(args):
    with patch("sys.argv", ["pyright-polite", *args]):
        result = parse_cli()
        assert result.argv[1:] == [*args]
        assert result.mode == Mode.JSON


def test_files():
    with patch("sys.argv", ["pyright-polite", "foo.py", "bar.py"]):
        result = parse_cli()
        assert result.argv[1:] == ["--outputjson", "foo.py", "bar.py"]
        assert result.mode == Mode.JSON


def test_bad_arg():
    with patch("sys.argv", ["pyright-polite", "--bad-arg"]):
        with pytest.raises(SystemExit) as pytest_e:
            parse_cli()

        assert pytest_e.type == SystemExit
        assert pytest_e.value.code == 4


@pytest.mark.disable_autouse()
def test_pyright_found():
    patcher_shutil = patch("shutil.which", return_value="pyright")
    patcher_shutil.start()

    patcher_access = patch("os.access", return_value=True)
    patcher_access.start()

    with patch("sys.argv", ["pyright-polite"]):
        parse_cli()

    patcher_access.stop()
    patcher_shutil.stop()


@pytest.mark.disable_autouse()
@pytest.mark.parametrize(
    "expected_returncode",
    [
        pytest.param(
            127,
            marks=pytest.mark.skipif(sys.platform == "win32", reason="POSIX-only"),
        ),
        pytest.param(
            9009,
            marks=pytest.mark.skipif(sys.platform != "win32", reason="Windows-only"),
        ),
    ],
)
def test_pyright_not_found(expected_returncode):
    patcher_shutil = patch("shutil.which", return_value=None)
    patcher_shutil.start()

    with patch("sys.argv", ["pyright-polite"]):
        with pytest.raises(SystemExit) as pytest_e:
            parse_cli()
        assert pytest_e.type == SystemExit
        assert pytest_e.value.code == expected_returncode

    patcher_shutil.stop()


@pytest.mark.disable_autouse()
@pytest.mark.skipif(sys.platform == "win32", reason="POSIX-only")
def test_pyright_not_executable():
    patcher_shutil = patch("shutil.which", return_value="pyright")
    patcher_shutil.start()

    patcher_access = patch("os.access", return_value=False)
    patcher_access.start()

    with patch("sys.argv", ["pyright-polite"]):
        with pytest.raises(SystemExit) as pytest_e:
            parse_cli()
        assert pytest_e.type == SystemExit
        assert pytest_e.value.code == 126

    patcher_access.stop()
    patcher_shutil.stop()


tests/test_main.py
# SPDX-FileCopyrightText: Copyright © 2023 Jamie Nguyen <j@jamielinux.com>
# SPDX-License-Identifier: MIT

import sys

import pytest


def test_main_entrypoint(event_loop, utils, capsys, stderr, stdout, expected_stdout):
    utils.mock_subproc(
        event_loop,
        capsys,
        subproc_stderr=stderr,
        subproc_stdout=stdout,
        expected_stdout=expected_stdout,
    )


@pytest.mark.parametrize(
    ("subproc_returncode", "expected_returncode"),
    [
        (1, 1),
        (None, 0),
    ],
)
def test_main_entrypoint_returncodes(
    event_loop,
    utils,
    capsys,
    stderr,
    stdout,
    expected_stdout,
    subproc_returncode,
    expected_returncode,
):
    utils.mock_subproc(
        event_loop,
        capsys,
        subproc_stderr=stderr,
        subproc_stdout=stdout,
        subproc_returncode=subproc_returncode,
        expected_stdout=expected_stdout,
        expected_returncode=expected_returncode,
    )


@pytest.mark.skipif(sys.platform == "win32", reason="POSIX-only")
def test_main_entrypoint_negative_returncode(
    event_loop, utils, capsys, stderr, stdout, expected_stdout
):
    utils.mock_subproc(
        event_loop,
        capsys,
        subproc_stderr=stderr,
        subproc_stdout=stdout,
        subproc_returncode=-1,
        expected_stdout=expected_stdout,
        expected_returncode=0,
    )


src/pyright_polite/main.py
# SPDX-FileCopyrightText: Copyright © 2023 Jamie Nguyen <j@jamielinux.com>
# SPDX-License-Identifier: MIT
# pyright: strict

"""Main entrypoint."""

import asyncio
import contextlib
import sys

from .cli import CommandLine, parse_cli
from .platform import WINDOWS
from .polite import Polite

__all__ = ["main"]


def main() -> None:
    """Main entrypoint."""
    cli: CommandLine = parse_cli()
    polite: Polite = Polite(cli)

    if WINDOWS:
        # The default event loop implementation for Windows changed in Python 3.8, so
        # explicitly pick the Proactor event loop.
        asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())  # type: ignore

    with contextlib.suppress(KeyboardInterrupt):
        returncode: int = asyncio.run(polite.run_pyright())
        sys.exit(returncode)


if __name__ == "__main__":
    main()


src/pyright_polite/polite.py
# SPDX-FileCopyrightText: Copyright © 2023 Jamie Nguyen <j@jamielinux.com>
# SPDX-License-Identifier: MIT
# pyright: strict
# pylint: disable=no-member  # (pylint bug)

"""The Polite class for launching and handling pyright as a subprocess."""

import asyncio
import contextlib
import signal
import sys
from enum import Enum, auto
from types import FrameType
from typing import Dict, List, Optional

from .cli import CommandLine, Mode
from .platform import STATUS_CONTROL_C_EXIT, TERM_SIGNAL, WINDOWS
from .pyright import Pyright, PyrightJsonError, parse_pyright, print_filtered, print_pyright

__all__ = ["Polite"]


class StreamName(Enum):
    """Output stream names."""

    STDERR = auto()
    STDOUT = auto()


class Polite:
    """Launch pyright, handle its output and react to signals."""

    __slots__ = (
        "subproc",
        "_returncode",
        "cli",
        "stdout_lines",
        "lock",
        "signalled",
        "signal_returncodes",
    )

    def __init__(self, cli: CommandLine) -> None:
        """Prepare the instance."""
        self.subproc: asyncio.subprocess.Process
        self._returncode: int = 0

        self.cli: CommandLine = cli
        self.stdout_lines: List[str] = []
        self.lock: asyncio.Lock = asyncio.Lock()
        self.signalled: bool = False

        self.signal_returncodes: Dict[signal.Signals, int]
        if WINDOWS:
            self.signal_returncodes = {
                signal.SIGINT: STATUS_CONTROL_C_EXIT,
                signal.SIGBREAK: STATUS_CONTROL_C_EXIT,  # type: ignore
            }
        else:
            self.signal_returncodes = {
                signal.SIGHUP: 129,  # type: ignore
                signal.SIGINT: 130,
                signal.SIGQUIT: 131,  # type: ignore
                signal.SIGALRM: 142,  # type: ignore
                signal.SIGTERM: 143,
            }

    @property
    def returncode(self) -> int:
        """Getter for the return code."""
        return self._returncode

    @returncode.setter
    def returncode(self, value: Optional[int]) -> None:
        """Setter for the return code."""
        if not isinstance(value, int):
            return

        # When using `asyncio.create_subprocess_exec()`, if the subprocess is killed
        # externally with a signal then the `returncode` property for the asyncio
        # subprocess will be negative. On Linux / Darwin, returncodes must be positive,
        # so we can't just exit using the subprocess returncode without checking.
        if value < 0 and not WINDOWS:
            return

        self._returncode = value

    def unset_signal_handlers(self) -> None:
        """Unset signal handlers back to default.

        For some reason on Windows ctrl-c stops working in powershell sessions unless
        we reset signal handlers back to default (ie, `SIG_DFL`).
        """
        if WINDOWS:
            for sig in self.signal_returncodes:
                signal.signal(sig, signal.SIG_DFL)

    def signal_handler(self, signum: int, _: Optional[FrameType]) -> None:
        """Signal handler to kill the pyright subprocess gracefully.

        Args:
            signum (int): The signal number.
        """
        # Using `loop.add_signal_handler()` isn't possible on Windows. Fortunately,
        # this signal handler doesn't need to do any loop cleanup itself, so we can
        # happily just use a synchronous signal handler.

        self.unset_signal_handlers()
        self.returncode = self.signal_returncodes[signal.Signals(signum)]
        self.signalled = True

        if self.subproc.returncode is None:
            with contextlib.suppress(ProcessLookupError):
                self.subproc.send_signal(TERM_SIGNAL)

    async def kill_subproc(self) -> None:
        """Kill the pyright subprocess with the appropriate signal."""
        if self.subproc.returncode is None:
            with contextlib.suppress(ProcessLookupError):
                self.subproc.send_signal(TERM_SIGNAL)
        await self.subproc.wait()

    async def just_wait_for_pyright(self) -> int:
        """Just wait for pyright to exit, without doing any output filtering.

        Args:
            int: The return code we want to exit with.
        """
        with contextlib.suppress(asyncio.CancelledError):
            await self.subproc.wait()
            self.unset_signal_handlers()
            if isinstance(self.subproc.returncode, int) and self.subproc.returncode > 0:
                return self.subproc.returncode
        return 0

    async def process_bytes(self, line_bytes: bytes, stream_name: StreamName) -> None:
        """Process bytes from an output stream.

        Args:
            line_bytes (bytes): The line retrieved by `readline()`.
            stream_name (StreamName): The name of the stream.
        """
        await asyncio.sleep(0)

        line = line_bytes.decode("utf-8")
        if stream_name == StreamName.STDERR or self.cli.mode == Mode.PLAINTEXT:
            print_filtered(line)
            return

        self.stdout_lines.append(line)
        if line.startswith("}"):
            pyright: Pyright = parse_pyright("".join(self.stdout_lines))
            async with self.lock:
                print_pyright(pyright)
            self.stdout_lines.clear()

    async def stream_handler(self, stream: asyncio.StreamReader, stream_name: StreamName) -> None:
        """Handle stdout and stderr streams.

        Args:
            stream (asyncio.StreamReader): The stream to handle.
            stream_name (StreamName): The name of the stream.
        """
        # At startup, try to print any initial stderr messages first like pyright does.
        # Even with an empty dir, pyright takes >=0.3s to print to stdout anyway.
        if stream_name == StreamName.STDOUT:
            await asyncio.sleep(0.3)

        while not stream.at_eof():
            line_bytes: bytes = await stream.readline()
            if line_bytes:
                # If using pyright from pypi, sending ctrl-c in `--watch` mode will
                # always result in a (completely useless) KeyboardInterrupt traceback.
                if (
                    stream_name == StreamName.STDERR
                    and self.signalled
                    and line_bytes.startswith(b"Traceback (most recent call last):")
                ):
                    return

                await self.process_bytes(line_bytes, stream_name)

        # EOF before a closing "}" means partial JSON output.
        if stream_name == StreamName.STDOUT and self.stdout_lines:
            raise PyrightJsonError("unexpected EOF")

    async def run_pyright(self) -> int:
        """Run pyright and filter and parse its output.

        Returns:
            int: The return code we want to exit with.
        """
        # pyright's `--dependencies` and `--stats` args both include diagnostics in their
        # output, but pyright forbids `--outputjson` with these so we can't parse easily.
        # If we skip parsing and just filter noise from the output, we'd still lose colors.
        # So just avoid capturing stderr/stdout altogether for these args.
        capture: bool = self.cli.mode != Mode.UNFILTERED

        self.subproc = await asyncio.create_subprocess_exec(
            *self.cli.argv,
            stdout=asyncio.subprocess.PIPE if capture else None,
            stderr=asyncio.subprocess.PIPE if capture else None,
        )

        for sig in self.signal_returncodes:
            signal.signal(sig, self.signal_handler)

        if (
            self.cli.mode == Mode.UNFILTERED
            or self.subproc.stdout is None
            or self.subproc.stderr is None
        ):
            self.returncode = await self.just_wait_for_pyright()

        else:
            tasks = [
                asyncio.create_task(self.stream_handler(self.subproc.stderr, StreamName.STDERR)),
                asyncio.create_task(self.stream_handler(self.subproc.stdout, StreamName.STDOUT)),
            ]

            # Possible exit case scenarios:
            #
            # 1. pyright exited normally (usually anything except `--watch` mode).
            # 2. pyright was killed by a signal (usually ctrl-c in `--watch` mode).
            # 3. pyright returned bad JSON.
            # 4. pyright closed stdout and stderr file descriptors, but is still running.
            #        This is unlikely, but possible if there's a bug in pyright.
            # 5. pyright-polite received a signal.

            with contextlib.suppress(asyncio.CancelledError):
                try:
                    # Cases 1 and 2.
                    await asyncio.gather(*tasks)
                    await self.subproc.wait()
                except PyrightJsonError as exc:
                    # Case 3.
                    sys.stderr.write(f"{exc}\n")
                    self.returncode = 2
                finally:
                    # Cases 3, 4 and 5.
                    self.unset_signal_handlers()
                    await self.kill_subproc()
                    for task in tasks:
                        if not task.done():
                            task.cancel()
                    await asyncio.gather(*tasks, return_exceptions=True)

            # We want to return the same exit code that pyright did, except:
            #
            # A. When pyright-polite receives a signal (ie, case 5), in which case we should
            #        use the appropriate exit codes for the signal received.
            # B. When the JSON is bad (ie, case 3), which should never happen really unless
            #        there's a bug in pyright or they changed their JSON schema.
            #
            # In both Case A and Case B, self.returncode would have been set to a
            # non-zero value by the time we get here.

            if self.returncode == 0:
                self.returncode = self.subproc.returncode

        return self.returncode


src/pyright_polite/platform.py
# SPDX-FileCopyrightText: Copyright © 2023 Jamie Nguyen <j@jamielinux.com>
# SPDX-License-Identifier: MIT
# pyright: strict
# pylint: disable=no-member

"""Platform-specific global variables."""

import platform
import signal

__all__ = ["LINUX", "DARWIN", "WINDOWS", "STATUS_CONTROL_C_EXIT", "TERM_SIGNAL"]

# `platform.system()` has the most thorough heuristics for determining the OS. It tries
# `os.uname()`, `sys.platform` and OS-specific calls in that order (stopping when it gets
# an answer), and then normalizes into categories that include Linux, Darwin, Windows.
PLATFORM: str = platform.system()
LINUX: bool = PLATFORM == "Linux"
DARWIN: bool = PLATFORM == "Darwin"
WINDOWS: bool = PLATFORM == "Windows"

# On Windows, pyright returns -1073741510 exit code when it receives CTRL_C_EVENT or
# CTRL_BREAK_EVENT. This number is the signed int representation of 0xC000013A, which is
# the general Windows error code indicating that a program was terminated with Ctrl-C.
STATUS_CONTROL_C_EXIT: int = -1073741510

# Send this signal to pyright when we want it to exit. Some people install pypi-pyright,
# which runs npm-pyright as a subprocess; importantly, for that scenario we must send a
# SIGINT (Linux/Darwin) or CTRL_C_EVENT (Windows) to pypi-pyright, instead of SIGTERM
# or anything else, otherwise it will leave npm-pyright running. The same signals are
# also appropriate when working directly with npm-pyright.
TERM_SIGNAL: signal.Signals = signal.CTRL_C_EVENT if WINDOWS else signal.SIGINT  # type: ignore


src/pyright_polite/cli.py
# SPDX-FileCopyrightText: Copyright © 2023 Jamie Nguyen <j@jamielinux.com>
# SPDX-License-Identifier: MIT
# pyright: strict

"""Handle command-line arguments."""

import argparse
import dataclasses
import os
import shutil
import sys
from enum import Enum, auto
from typing import List, Optional, Set, Union

from .platform import WINDOWS
from .pyright import Severity

__all__ = ["parse_cli"]

HELP: str = """\
Usage: pyright-polite [options] files...
  Options:
  --createstub <IMPORT>              Create type stub file(s) for import
  --dependencies                     Emit import dependency information
  -h,--help                          Show this help message
  --ignoreexternal                   Ignore external imports for --verifytypes
  --lib                              Use library code to infer types when stubs are missing
  --level <LEVEL>                    Minimum diagnostic level (error or warning)
  --outputjson                       Output results in JSON format
  -p,--project <FILE OR DIRECTORY>   Use the configuration file at this location
  --pythonplatform <PLATFORM>        Analyze for a specific platform (Darwin, Linux, Windows)
  --pythonversion <VERSION>          Analyze for a specific version (3.3, 3.4, etc.)
  --skipunannotated                  Skip analysis of functions with no type annotations
  --stats                            Print detailed performance stats
  -t,--typeshed-path <DIRECTORY>     Use typeshed type stubs at this location
  -v,--venv-path <DIRECTORY>         Directory that contains virtual environments
  --verbose                          Emit verbose diagnostics
  --verifytypes <PACKAGE>            Verify type completeness of a py.typed package
  --version                          Print Pyright version
  --warnings                         Use exit code of 1 if warnings are reported
  -w,--watch                         Continue to run and watch for changes


  Note: pyright-polite does not filter output from `--dependencies` or `--stats`.

"""


UNFILTERED_ARGS: Set[str] = {
    "--dependencies",
    "--stats",
    "--version",
}


PLAINTEXT_ARGS: Set[str] = {
    "--createstub",
    "--ignoreexternal",
    "--verifytypes",
}

NON_JSON_ARGS: Set[str] = PLAINTEXT_ARGS | UNFILTERED_ARGS


class ArgumentParser(argparse.ArgumentParser):
    """A subclass of `argparse.ArgumentParser` just to change the returncode."""

    def error(self, message: str):
        """Custom `error` method to exit with a returncode of 4 instead of 2.

        Pyright exits with a returncode of 4 when launched with invalid command-line
        arguments, so we should do the same.

        The `exit_on_error` parameter for `argparse.ArgumentParser` was only introduced
        in Python 3.9, so we can't make use of that.
        """
        self.print_usage(sys.stderr)
        args = {"prog": self.prog, "message": message}
        self.exit(4, ("%(prog)s: error: %(message)s\n") % args)


class CustomHelpFormatter(argparse.HelpFormatter):
    """Custom help formatter to display exactly the same help text as pyright."""

    def format_help(self) -> str:
        """Print help message."""
        return HELP


class Mode(Enum):
    """The mode of operation for pyright-polite."""

    # Filter stderr. Parse JSON output in stdout.
    JSON = auto()

    # Filter both stderr and stdout. No JSON expected.
    PLAINTEXT = auto()

    # Don't capture stderr/stdout (ie, let pyright write directly to the terminal).
    UNFILTERED = auto()


@dataclasses.dataclass(frozen=True)
class CommandLine:
    """The argv list and the deduced mode of operation."""

    argv: List[str]
    mode: Mode


def find_pyright() -> str:
    """Find the pyright executable.

    `asyncio.create_subprocess_exec()` requires the full path on Windows, and we want to
    avoid using `asyncio.create_subprocess_shell()`.

    Returns:
        str: The fully-qualified absolute path to the pyright executable.
    """
    # The default mode for shutil.which() is `F_OK | X_OK`, but to give a more helpful
    # error message we'll check X_OK in a separate step.
    pyright: Optional[str] = shutil.which("pyright", mode=os.F_OK)

    if pyright is None:
        sys.stderr.write("pyright-polite: error: pyright could not be found in your PATH\n")
        returncode = 9009 if WINDOWS else 127
        sys.exit(returncode)
    if not os.access(pyright, os.X_OK):  # always True on Windows
        sys.stderr.write("pyright-polite: error: pyright is not executable\n")
        sys.exit(126)

    return pyright


def prepare_argv(parsed_args: argparse.Namespace) -> List[str]:
    """Prepare argv for `asyncio.create_subprocess_exec()`.

    This includes the pyright executable (ie, argv[0]).

    pyright has args that are incompatible with each other, but instead of recreating
    pyright's compatibility logic we'll just let pyright complain if it receives
    incompatible args.

    Args:
        parsed_args (argparse.Namespace): The parsed arguments.

    Returns:
        List[str]: The argv list to pass to `asyncio.create_subprocess_exec()`.
    """
    pyright: str = find_pyright()
    argv: List[str] = [pyright]

    # If --version is passed anywhere on the command-line, ignore all other arguments
    # and just skip to executing `pyright --version`.
    if parsed_args.version:
        argv.append("--version")
        return argv

    arg_mapping = {
        "createstub": "--createstub",
        "dependencies": "--dependencies",
        "ignoreexternal": "--ignoreexternal",
        "lib": "--lib",
        "level": "--level",
        "project": "--project",
        "pythonplatform": "--pythonplatform",
        "pythonversion": "--pythonversion",
        "skipunannotated": "--skipunannotated",
        "stats": "--stats",
        "typeshed_path": "--typeshed-path",
        "venv_path": "--venv-path",
        "warnings": "--warnings",
        "verbose": "--verbose",
        "verifytypes": "--verifytypes",
        "watch": "--watch",
    }

    for arg_property, flag in arg_mapping.items():
        arg_value: Union[bool, List[str], None] = getattr(parsed_args, arg_property)
        if arg_value:
            argv.append(flag)
            if isinstance(arg_value, list):
                argv.append(arg_value[0])

    # If the user passes `--outputjson` manually, always pass this to pyright even if
    # it will clash with one of the other command-line arguments that don't support
    # this mode. We'll just let pyright complain to the user about the clash.
    if parsed_args.outputjson or not any(arg in NON_JSON_ARGS for arg in argv[1:]):
        argv.append("--outputjson")

    for file in parsed_args.files:
        argv.append(file)

    return argv


def deduce_mode(args: List[str]) -> Mode:
    """Deduce the mode of operation based on the arguments passed to pyright.

    Returns:
        Mode: The deduced mode of operation.
    """
    if "--outputjson" in args:
        return Mode.JSON

    if any(arg in UNFILTERED_ARGS for arg in args):
        return Mode.UNFILTERED

    return Mode.PLAINTEXT


def parse_cli() -> CommandLine:
    """Handle command-line arguments.

    Returns:
        CommandLine: The argv list and the deduced mode of operation.
    """
    parser: argparse.ArgumentParser = ArgumentParser(
        formatter_class=CustomHelpFormatter,
        allow_abbrev=False,
    )
    parser.add_argument("--createstub", nargs=1, type=str)
    parser.add_argument("--dependencies", action="store_true")
    parser.add_argument("--ignoreexternal", action="store_true")
    parser.add_argument("--lib", action="store_true")
    parser.add_argument("--level", nargs=1, choices=[s.name.lower() for s in Severity])
    parser.add_argument("--outputjson", action="store_true")
    parser.add_argument("-p", "--project", nargs=1, type=str)
    parser.add_argument("--pythonplatform", nargs=1, type=str)
    parser.add_argument("--pythonversion", nargs=1, type=str)
    parser.add_argument("--skipunannotated", action="store_true")
    parser.add_argument("--stats", action="store_true")
    parser.add_argument("-t", "--typeshed-path", nargs=1, type=str)
    parser.add_argument("-v", "--venv-path", nargs=1, type=str)
    parser.add_argument("--warnings", action="store_true")
    parser.add_argument("--verbose", action="store_true")
    parser.add_argument("--verifytypes", nargs=1, type=str)
    parser.add_argument("--version", action="store_true")
    parser.add_argument("-w", "--watch", action="store_true")
    parser.add_argument("files", nargs="*")

    parsed_args: argparse.Namespace = parser.parse_args()
    argv: List[str] = prepare_argv(parsed_args)
    mode: Mode = deduce_mode(argv[1:])
    return CommandLine(argv=argv, mode=mode)
